Adds security enhancements

userhas404d opened the pull request:
## Security Group and key pair handling
- updates main module to accept only the key pair name 
- pushes the key pair handling to caller module(s)
- updates examples to conditionally create the ssh key pair and conditionally write the private key to disk
- updates runner security group creation to conditionally enable inbound ssh connections if a key pair name is specified
- adds the caller's public IP address to inbound ssh rule if a key pair is specified, 
- adds the option to specify additional inbound addresses in CIDR notation
## Misc changes
- various spelling fixes and some opinionated grammatical adjustments to variable descriptions
- updates terraform version and module version references
- updates the readme

userhas404d commented:
Just wanted to say thanks for all your work on this awesome module! Sorry if the commit history is a bit weird here - I rebased this branch against develop in order to test with the changes from the latest release. I can roll that back if that throws things too far out of whack.

npalm commented:
@userhas404d thanks for the PR. Will do my best to check the PR in the next days.

npalm commented on examples/runner-public/key.tf:
```diff
@@ -1,25 +1,38 @@
+  key_pair_write          = "${local.key_pair_specified && var.write_private_key}"
+  generated_key_pair_name = "${var.environment}-gitlab-runner"
+  key_pair_name           = "${ local.key_pair_specified ? local.generated_key_pair_name : var.key_pair_name }"
+  key_pair_path           = "${path.module}/generated}"
```
Last `}` should not be there

userhas404d replied:
rebased against your other fixes so this should be good now

npalm commented on examples/runner-public/main.tf:
```diff
@@ -19,13 +19,14 @@ module "runner" {
   environment = "${var.environment}"
 
-  ssh_public_key = "${local_file.public_ssh_key.content}"
+  ssh_key_name = "${local.key_pair_name}"
```
Should refer the the terraform resource not to the local var to avoid this resource (launch group) is created before the keypair

userhas404d replied:
I think this is good now thanks to your other commit(s)

npalm commented:
@userhas404d Made some changes, pushed the updated branch  to https://github.com/npalm/terraform-aws-gitlab-runner/tree/userhas404d-lockdown The example `runners-public` is now working.

userhas404d commented on main.tf:
```diff
@@ -1,17 +1,26 @@
+    to_port         = 22
+    protocol        = "tcp"
+    cidr_blocks     = ["${local.cidr_blocks_allowed_inbound}"]
+    security_groups = ["${var.allow_ssh_to_runner_instance_sg}"]
```
Oops! you're right, I was testing this against my env and overlooked actually running the included examples.. I ended up splitting the CIDR and security group rules that allow SSH inbound to fix this.

userhas404d commented on main.tf:
```diff
@@ -1,17 +1,26 @@
+    from_port       = 22
+    to_port         = 22
+    protocol        = "tcp"
+    cidr_blocks     = ["${local.cidr_blocks_allowed_inbound}"]
```
1. So my reasoning here is that ideally, the runner instances should allow outbound traffic _only_. The instances themselves will still be able to make outbound requests and pull in data from the rest of the internet. I'm just limiting any outside connections being initiated here. That's actually how I have this module deployed in my env currently.

2. completely agree, I went ahead and pushed the ssh inbound availability completely up to the user calling the module.

userhas404d commented on main.tf:
```diff
@@ -24,43 +33,54 @@ resource "aws_security_group" "runner" {
   type        = "egress"
   from_port   = 0
-  to_port     = 65535
+  to_port     = 0
```
Tracking this across the AWS api documentation was harder than I thought it was going to be so I threw this together.

```hcl
data "aws_vpc" "selected" {
  id = "YOUR_VPC"
}

resource "aws_security_group" "test_out_1" {
  name   = "test_out_1"
  vpc_id = "${data.aws_vpc.selected.id}"
}

resource "aws_security_group_rule" "test_out_1_rule" {
  type              = "egress"
  to_port           = 0
  protocol          = "-1"
  cidr_blocks       = ["0.0.0.0/0"]
  from_port         = 65535
  security_group_id = "${aws_security_group.test_out_1.id}"
}

resource "aws_security_group" "test_out_2" {
  name   = "test_out_2"
  vpc_id = "${data.aws_vpc.selected.id}"

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}

output "egress_out_1" {
  depends_on = ["aws_security_group_rule.test_out_1_rule"]
  value      = "${aws_security_group.test_out_1.egress}"
}

output "egress_out_2" {
  value = "${aws_security_group.test_out_2.egress}"
}
```

```hcl
Outputs:

egress_out_1 = [
    {
        cidr_blocks = [0.0.0.0/0],
        description = ,
        from_port = 0,
        ipv6_cidr_blocks = [],
        prefix_list_ids = [],
        protocol = -1,
        security_groups = [],
        self = 0,
        to_port = 0
    }
]
egress_out_2 = [
    {
        cidr_blocks = [0.0.0.0/0],
        description = ,
        from_port = 0,
        ipv6_cidr_blocks = [],
        prefix_list_ids = [],
        protocol = -1,
        security_groups = [],
        self = 0,
        to_port = 0
    }
]
```

Alternatively, running this:

```hcl
resource "aws_security_group" "test_out_1" {
  name   = "test_out_1"
  vpc_id = "${data.aws_vpc.selected.id}"

  egress {
    from_port   = 0
    to_port     = 65535
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}
```

Results in the following error:

```hcl
Error: Error applying plan:

1 error(s) occurred:

* aws_security_group.test_out_1: 1 error(s) occurred:

* aws_security_group.test_out_1: from_port (0) and to_port (65535) must both be 0 to use the 'ALL' "-1" protocol!
```

userhas404d commented on main.tf:
```diff
@@ -102,11 +122,17 @@ data "template_file" "gitlab_runner" {
 }
 
+# validate that the specified runner subnet is in the target AZ
+data "aws_subnet" "runners" {
```
Interesting, I need to test this more then..

userhas404d commented:
@npalm I rebased against develop and the target branch but pretty sure I introduced some regressions along the way.. I'll ping you again once I have this ready for another round of review.

npalm commented:
@userhas404d I saw your original PR was against the branch for the auto registration. I am still testing that one. Was it working for you? I hope I can finish the work on that branch in the next days. 

Thanks once again for sharing!

userhas404d commented:
So I think this was the issue for auto registration. Had to step away for a bit (on mobile) so hopefully this link is the correct one: https://github.com/npalm/terraform-aws-gitlab-runner/commit/6e19a26708ea0c26280041e356d19d3808617187

userhas404d commented:
@npalm think I bit off more than I could chew with the number of changes I've made and how quickly develop is moving. I'm going to go ahead and close this PR and open some smaller, more manageable ones when I have time.

npalm commented:
@userhas404d I really appreciate your work to improve the security. I lets see if we can do it in smaller chucks. Why not open a few issues to define the enhancement. Maybe we can do some joint effort. I also have no big changes ahead / planned. But the work on the auto registration branch was already going on for some time. I really want the feature back. So feel free to create a few issues or PR's.
